### PR TITLE
refactor to use Item class

### PR DESCRIPTION
**Pull request by ThomasWaldmann**
squash / merge?

### CODE REVIEW

File: src/borg/archive.py [hunk 1]
```diff
@@ -86,7 +87,7 @@ def show_progress(self, item=None, final=False, stream=None, dt=None):
             columns, lines = get_terminal_size()
             if not final:
                 msg = '{0.osize_fmt} O {0.csize_fmt} C {0.usize_fmt} D {0.nfiles} N '.format(self)
```
Comment (RonnyPfannschmidt) [inline src/borg/archive.py]: i think this would be more awesome, if __format__ was implemented, and the attribute names where combined with getattr

Comment (ThomasWaldmann) [inline src/borg/archive.py]: out of scope for this PR, but misc. improvements can be done later.

File: src/borg/archive.py [hunk 2]
```diff
@@ -106,16 +107,16 @@ def unpack_many(self, ids, filter=None, preload=False):
         unpacker = msgpack.Unpacker(use_list=False)
         for _, data in self.fetch_many(ids):
             unpacker.feed(data)
-            items = [decode_dict(item, ITEM_TEXT_KEYS) for item in unpacker]
+            items = [Item(internal_dict=item) for item in unpacker]
```
Comment (ThomasWaldmann) [inline src/borg/archive.py]: the internal dict of Item is holding data in the format that can be given to msgpack (and thus is also fed with data received from msgpack).

File: src/borg/archive.py [hunk 3]
```diff
@@ -135,7 +136,7 @@ def __init__(self, key, chunker_params=ITEMS_CHUNKER_PARAMS):
         self.chunker = Chunker(self.key.chunk_seed, *chunker_params)
 
     def add(self, item):
-        self.buffer.write(self.packer.pack(StableDict(item)))
+        self.buffer.write(self.packer.pack(item.as_dict()))
```
Comment (ThomasWaldmann) [inline src/borg/archive.py]: == give data from internal dict to msgpack

File: src/borg/archive.py [hunk 4]
```diff
@@ -698,7 +693,7 @@ class RobustUnpacker:
     """
     def __init__(self, validator):
         super().__init__()
-        self.item_keys = [msgpack.packb(name) for name in ITEM_KEYS]
+        self.item_keys = [msgpack.packb(name.encode()) for name in ITEM_KEYS]
```
Comment (ThomasWaldmann) [inline src/borg/archive.py]: change element type of ITEM_KEYS to str, so need to encode here.

File: src/borg/archive.py [hunk 5]
```diff
@@ -894,10 +889,10 @@ def verify_file_chunks(item):
             """
             offset = 0
             chunk_list = []
-            for chunk_id, size, csize in item[b'chunks']:
+            for chunk_id, size, csize in item.chunks:
                 if chunk_id not in self.chunks:
                     # If a file chunk is missing, create an all empty replacement chunk
-                    logger.error('{}: Missing file chunk detected (Byte {}-{})'.format(safe_decode(item[b'path']), offset, offset + size))
+                    logger.error('{}: Missing file chunk detected (Byte {}-{})'.format(item.path, offset, offset + size))
```
Comment (ThomasWaldmann) [inline src/borg/archive.py]: getattr path does safe_decode (setattr does safe_encode).

File: src/borg/fuse.py [hunk 6]
```diff
@@ -164,45 +164,45 @@ def getattr(self, inode, ctx=None):
...
             else:
-                entry.st_atime = bigint_to_int(item[b'mtime']) / 1e9
-            if b'ctime' in item:
-                entry.st_ctime = bigint_to_int(item[b'ctime']) / 1e9
+                entry.st_atime = item.mtime / 1e9
+            if 'ctime' in item:
+                entry.st_ctime = item.ctime / 1e9
             else:
-                entry.st_ctime = bigint_to_int(item[b'mtime']) / 1e9
+                entry.st_ctime = item.mtime / 1e9
```
Comment (ThomasWaldmann) [inline src/borg/fuse.py]: this if/else block can be optimized later. kept it close to original within this changeset.

File: src/borg/item.py [hunk 7]
```diff
@@ -21,25 +21,34 @@ class PropDict:
...
-                raise TypeError("dict keys must be str or bytes, not %r" % k)
-            _dict[k] = v
-        unknown_keys = set(_dict) - self.VALID_KEYS
-        if unknown_keys:
-            raise ValueError("dict contains unknown keys %s" % ','.join(unknown_keys))
-        self._dict = _dict
+            self._dict[k] = v
+
+    def __eq__(self, other):
+        return self.as_dict() == other.as_dict()
```
Comment (ThomasWaldmann) [inline src/borg/item.py]: this was a try to make unittests self.assertEqual() happy, but might need more work.

File: src/borg/item.py [hunk 8]
```diff
@@ -110,22 +119,22 @@ class Item(PropDict):
     If an Item shall be serialized, give as_dict() method output to msgpack packer.
     """
 
-    VALID_KEYS = set(key.decode() for key in ITEM_KEYS)  # we want str-typed keys
+    VALID_KEYS = ITEM_KEYS | {'deleted'}  # str-typed keys
```
Comment (ThomasWaldmann) [inline src/borg/item.py]: "deleted" is a key used by "borg diff" internally but not written into the repo.

File: src/borg/item.py [hunk 9]
```diff
@@ -110,22 +119,22 @@ class Item(PropDict):
...
 
     __slots__ = ("_dict", )  # avoid setting attributes not supported by properties
 
     # properties statically defined, so that IDEs can know their names:
 
     path = PropDict._make_property('path', str, 'surrogate-escaped str', encode=safe_encode, decode=safe_decode)
     source = PropDict._make_property('source', str, 'surrogate-escaped str', encode=safe_encode, decode=safe_decode)
+    user = PropDict._make_property('user', (str, type(None)), 'surrogate-escaped str or None', encode=safe_encode, decode=safe_decode)
+    group = PropDict._make_property('group', (str, type(None)), 'surrogate-escaped str or None', encode=safe_encode, decode=safe_decode)
+
```
Comment (ThomasWaldmann) [inline src/borg/item.py]: above 4 attrs are the ones encoded/decoded in old code also.

File: src/borg/testsuite/archive.py [hunk 10]
```diff
@@ -93,7 +94,7 @@ def test_without_microseconds(self):
 class ChunkBufferTestCase(BaseTestCase):
 
     def test(self):
-        data = [{b'foo': 1}, {b'bar': 2}]
+        data = [Item(path='p1'), Item(path='p2')]
```
Comment (ThomasWaldmann) [inline src/borg/testsuite/archive.py]: had to change key name to make it pass the validation.

File: src/borg/testsuite/archive.py [hunk 11]
```diff
@@ -105,11 +106,11 @@ def test(self):
...
         for id in chunks.chunks:
             unpacker.feed(cache.objects[id])
-        self.assert_equal(data, list(unpacker))
+        self.assert_equal(data, [Item(internal_dict=d) for d in unpacker])
 
     def test_partial(self):
-        big = b"0123456789" * 10000
-        data = [{b'full': 1, b'data': big}, {b'partial': 2, b'data': big}]
+        big = "0123456789" * 10000
+        data = [Item(path='full', source=big), Item(path='partial', source=big)]
```
Comment (ThomasWaldmann) [inline src/borg/testsuite/archive.py]: had to change key name to make it pass the validation.

File: src/borg/archive.py [hunk 12]
```diff
@@ -286,9 +287,6 @@ def iter_items(self, filter=None, preload=False):
             yield item
 
     def add_item(self, item):
-        unknown_keys = set(item) - ITEM_KEYS
-        assert not unknown_keys, ('unknown item metadata keys detected, please update constants.ITEM_KEYS: %s',
```
Comment (ThomasWaldmann) [inline src/borg/archive.py]: this validation is now done in Item class

File: src/borg/platform/darwin.pyx [hunk 13]
```diff
@@ -62,28 +62,26 @@ def acl_get(path, item, st, numeric_owner=False):
...
+            item['acl_extended'] = text
     finally:
         acl_free(text)
         acl_free(acl)
 
 
 def acl_set(path, item, numeric_owner=False):
     cdef acl_t acl = NULL
-    try:
+    acl_text = item.get('acl_extended')
```
Comment (ThomasWaldmann) [inline src/borg/platform/darwin.pyx]: needed to rewrite this to use .get as there is no __getitem__ any more.
it's now more similar to other platform code we already have.

**Comment by ThomasWaldmann** [comment]
fuse still has an issue:
```
(borg-env) This-MacBook-Pro:borg vagrant$ ./borg.exe mount repo::1 mnt
Local Exception.
Traceback (most recent call last):
  File "borg/archiver.py", line 2162, in main
  File "borg/archiver.py", line 2103, in run
  File "borg/archiver.py", line 88, in wrapper
  File "borg/archiver.py", line 721, in do_mount
  File "borg/fuse.py", line 68, in __init__
  File "borg/fuse.py", line 108, in process_archive
  File "borg/fuse.py", line 41, in add
  File "msgpack/__init__.py", line 47, in packb
  File "msgpack/_packer.pyx", line 231, in msgpack._packer.Packer.pack (msgpack/_packer.cpp:231)
  File "msgpack/_packer.pyx", line 233, in msgpack._packer.Packer.pack (msgpack/_packer.cpp:233)
  File "msgpack/_packer.pyx", line 228, in msgpack._packer.Packer._pack (msgpack/_packer.cpp:228)
TypeError: can't serialize Item(internal_dict={'gid': 0, 'group': b'wheel', 'atime': 1464968932000000000, 'mode': 16893, 'ctime': 1464966479000000000, 'mtime': 1464966479000000000, 'user': b'vagrant', 'uid': 502})

Platform: Darwin This-MacBook-Pro.local 14.4.0 Darwin Kernel Version 14.4.0: Thu May 28 11:35:04 PDT 2015; root:xnu-2782.30.5~1/RELEASE_X86_64 x86_64 i386
Borg: 1.0.4.dev276+ng4896f1f  Python: CPython 3.5.1
PID: 61975  CWD: /vagrant/borg
sys.argv: ['./borg.exe', 'mount', 'repo::1', 'mnt']
SSH_ORIGINAL_COMMAND: None
```

**Comment by ThomasWaldmann** [comment]
I'm squashing / merging this now ...

**Pull request merged**